Block Supports: Add background image support to Group block

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes #14744, part of #16479. Previous attempt: https://github.com/WordPress/gutenberg/pull/39243

This PR explores adding a Background block support, with a single `backgroundImage` property to begin with. The Group block is opted-in to this support.

Note: this PR originally proposed a Media block support as the container for this sub-feature. The PR currently reflects a Background block support, and the final naming is not (yet) decided — this is currently in discussion in the comments.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

As discussed in #14744, it is useful for more container blocks to have background image support, rather than just the Cover block. The approach in this PR adds in background image support (within Background) as a generalised block support so that more blocks could use the feature in the future, too. Of note: the approach in this PR dynamically injects the background image and sizing at render time, which gives the block support flexibility, and means we don't need to worry about block deprecations.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

* Add a Background block support and panel to the block inspector
* Add a Background Image control and option for the background block support (based on the Media controls in the Site Logo block)
* Opt the Group block in to using this block support
* Dynamically render an inline style for the background image when an image is set, and default to a background size of `cover` (with the ability for us to customise this value in follow-ups)

## Not covered

* If the approach here becomes viable, then in follow-ups we could add in additional controls and properties for things like:
  * Background position (as in the Cover block)
  * Background repeat
  * Background size for repeating (to control how large or small repeated background images are)

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

1. In a blocks theme that uses `appearanceTools` (e.g. TT3) add a Group block to a post or page, with a few blocks within it, and maybe some padding.
2. In the block inspector at the right hand side, you should now see a Media panel.
3. Explore adding an image from the media library.
4. Save the post and load the site frontend, and ensure the background image renders.

## To-do

* [x] ~Fix skipSerialization (currently the styles _are_ being serialized when they shouldn't be)~

## Screenshots or screencast <!-- if applicable -->

<img width="1419" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/5e193ef3-6851-4800-a864-aeeaa13b4ec4">

---

<!--
https://github.com/WordPress/gutenberg/assets/14988353/c615d1ba-2731-432f-8c98-4db7f20c40ef
https://github.com/WordPress/gutenberg/assets/14988353/2f6493dc-b48c-4f7a-ab36-02b4a849e294
-->

The following screengrab demos adding a background image with a transparent background, and then selecting a background image that has no transparency:

https://github.com/WordPress/gutenberg/assets/14988353/82729e70-8a54-427b-b410-033e0a19faa9

andrewserong committed:
Media Block Support: Add backgroundImage support to Group block

andrewserong commented on packages/block-editor/src/hooks/style.js:
```diff
@@ -127,6 +129,7 @@ const skipSerializationPathsEdit = {
 const skipSerializationPathsSave = {
 	...skipSerializationPathsEdit,
 	[ `${ SPACING_SUPPORT_KEY }` ]: [ 'spacing.blockGap' ],
+	[ `${ MEDIA_SUPPORT_KEY }` ]: [ MEDIA_SUPPORT_KEY ],
```
This isn't quite working as intended yet — I notice the background styles are being serialized. I'll fix this up next week so that they're skipped for these styles.

andrewserong replied:
Fixed in https://github.com/WordPress/gutenberg/pull/53934/commits/546c0ad1073fd61d7b317abf994400f4953a8d31

scruffian commented on lib/block-supports/media.php:
```diff
@@ -0,0 +1,104 @@
+		return $block_content;
+	}
+
+	// TODO: should we get the style engine to handle `backgroundImage`?
```
That would make sense to me

andrewserong replied:
Great! I made a start on looking at adding it to the style engine but ran into a couple of edge cases. I might leave things here for now and split out a separate PR for looking at adding in backgroundImage / backgroundSize to the style engine.

ramonjd replied:
Out of interest, what were the edge cases?

andrewserong replied:
> Out of interest, what were the edge cases?

It turns out the edge cases were me doing things wrong 😄

I _think_ I've sorted it out now, I've pushed a commit in https://github.com/WordPress/gutenberg/pull/53934/commits/b836572cf141b78d0a72f7c71d3a9c2dd09dddb7 that seems to work for now. Since this PR is getting pretty big, though, if the approach in this PR winds up being viable, I can split out those changes in a separate PR to make it easier for review.

scruffian commented on lib/block-supports/media.php:
```diff
@@ -0,0 +1,104 @@
+
+	$background_image_source = _wp_array_get( $block_attributes, array( 'style', 'media', 'backgroundImage', 'source' ), null );
+	$background_image_url    = _wp_array_get( $block_attributes, array( 'style', 'media', 'backgroundImage', 'url' ), null );
+	$background_size         = _wp_array_get( $block_attributes, array( 'style', 'media', 'backgroundSize' ), null );
```
I think the last param here is the default value, so could you just make it `cover` and remove the lines below?

scruffian commented:
Once I have uploaded an image, if I try to change it I can open the media library and select a video, which doesn't work....

scruffian commented on packages/block-editor/src/hooks/media.scss:
```diff
@@ -0,0 +1,61 @@
+
+.block-editor-hooks__media__inspector-upload-container,
+.block-editor-hooks__media__inspector-media-replace-container {
+	button.components-button {
```
I feel like we should make this a variation of the button component rather than having a custom rule for it here. That should be a different PR though.

youknowriad commented:
- I've noticed that the styles are duplicated in the frontend.
- I like that the style engine (JS one) also supports this.
- If I'm not wrong, the cover block uses a separate div for this, so I wonder if it's possible to actually unify the cover with this behavior at some point (I prefer to avoid extra DOM nodes here)
- Any thoughts on global styles support for this?

youknowriad commented:
I can't seem to make both "color background" and "image background" work at the same time? what about with gradients?

jameskoster commented:
Excited to see this under way!

I know you mentioned it wont be covered in this PR, but in terms of design it would be good to understand the broader vision for this. Is the intention to replicate the Cover experience, which is quite prescriptive? Or should we aim for something more flexible with options like background-size/position/repeat (as suggested), stacked backgrounds (images and colors), and so on?

cc @WordPress/theme-team for thoughts.

draganescu commented:
I am definitely missing a lot of context and knowledge but as I have explored a supports icon idea in the Icon/Vector registration API #53510 discussion and the test plugin - [Demo of WordPress Icon API](https://github.com/draganescu/wp-icon-api/tree/trunk) - I think a `media` supports should not mean background styles.

Why is support media = background media? The block should decide what to do with the media. The support opts in into a standard way of picking media.

paaljoachim commented:
It would actually be great if we could create a standarized Media settings panel. 
For media, URL (Youtube, Vimeo etc) or something else.

Having a focal picker would also be nice. Having a Media settings panel for any block that can use media.
![Screenshot 2023-08-25 at 21 45 41](https://github.com/WordPress/gutenberg/assets/5323259/8cff4c55-07c1-4a8c-adab-506e9b1156a6)

I do see the Media settings evolve which can be used for various blocks.

andrewserong committed:
Add error handling for video uploads, refactor to a BackgroundImagePanelItem component

andrewserong committed:
Ensure styles are not serialized in save mode, and not duplicated on site frontend

andrewserong committed:
Fix upload issue

andrewserong commented:
Thanks for the early reviewing and testing, everyone! There were lots of comments, so I'll reply here, a little out of sequence. Apologies if I missed anything!

> I've noticed that the styles are duplicated in the frontend.

I've fixed this up — styles should only be output once, and they shouldn't be serialized to post content. This allows us to look at more dynamic behaviour in follow-ups such as setting background to Featured Image.

> Any thoughts on global styles support for this?

Global styles is an interesting case. I imagine blocks that could use it could include Post Content. But personally I'd be very interested in a root-global styles control for media/background image, particularly to set site-wide patterned backgrounds (like those in https://www.transparenttextures.com/). Since there would be lots of decisions to be made about how we'd implement it, I'm leaning toward us starting with just the block level for now, and build it out in global styles in follow-ups. This is another reason to keep all styles injection dynamic, so that we can change things in the future.

> I can't seem to make both "color background" and "image background" work at the same time? what about with gradients?

In contrast to Cover block, this block support doesn't do layering / overlays, so background color will only work with transparent images, but it does appear to work. There is a styling conflict with gradients, though, as gradients use the `background` property, which overrides `background-image`. If / when this PR winds up landing, I reckon adding in mixed gradient / background image support (via multiple values output within a single `background-image` property) would be good to explore. I'd leave mixed gradient + transparent background images to follow-ups, though. Here's a screengrab with a transparent image:

https://github.com/WordPress/gutenberg/assets/14988353/30a9407c-93fe-46be-8a2a-7a452614958e

> Or should we aim for something more flexible with options like background-size/position/repeat (as suggested), stacked backgrounds (images and colors), and so on?

My intention with this PR is for us to start out as simple as possible, and then add more advanced controls in follow-ups (potentially hidden by default in the ToolsPanel). I'd love to see control over background size within repeat, so that folks can scale up/down their repeating images as desired. But again, for follow-ups — would be happy to see some design explorations for how the current PR might scale, though!

> I am definitely missing a lot of context and knowledge but as I have explored a supports icon idea in the Icon/Vector registration API https://github.com/WordPress/gutenberg/discussions/53510 discussion and the test plugin - [Demo of WordPress Icon API](https://github.com/draganescu/wp-icon-api/tree/trunk) - I think a media supports should not mean background styles. Why is support media = background media?

Thanks for the feedback! The intention wasn't for `media === background media`, but rather to create a Media block support, and for the first feature within it, to add a background image support. In follow-ups, we might create all kinds of other media controls, but I think `media` could make for a good and consistent bucket for it. To make the hierarchy clearer, I've refactored the code a bit, so that the background image controls are in a separate sub-component, and I've updated the button label for clarity:

![image](https://github.com/WordPress/gutenberg/assets/14988353/49d25ddf-c845-491f-aa38-c5433ce3bea0)

> I do see the Media settings evolve which can be used for various blocks.

Thanks! Yes, my hope is that if we settle on Media as a panel, it could help with consistency across blocks, too 👍

> Once I have uploaded an image, if I try to change it I can open the media library and select a video, which doesn't work....

This _should_ be fixed now, and I've added a notice for if the media type is incorrect on upload. Let me know if you're still encountering it, though:

![image](https://github.com/WordPress/gutenberg/assets/14988353/cec4c66f-e7c0-4715-8b41-bc395c92c386)

---

Thanks again for all the feedback, folks! Let me know if there's other changes anyone would like to see. My objective for now is to try to focus on a decent MVP, and hopefully defer further features to follow-ups. I still have a little cleaning up to do with the PHP code, which I ran out of time to do today. Once that's done, and pending any design feedback / ideas, I'll switch this over to ready for review.

youknowriad commented:
> I've fixed this up — styles should only be output once, and they shouldn't be serialized to post content. This allows us to look at more dynamic behaviour in follow-ups such as setting background to Featured Image.

This is I think not the first place where we favor server rendering over post content serialization for styles. The downside of this approach is that it breaks these blocks for third-party block editors. These customization can in theory be easily supported for server less block editors but we don't have the functions that render the right HTML for that (including all these dynamic styles that are static in reality). It would be good to keep that in mind and maybe keep track of all these static/dynamic styles to implement the `wp.blocks.render` function properly.

andrewserong commented:
> This is I think not the first place where we favor server rendering over post content serialization for styles. The downside of this approach is that it breaks these blocks for third-party block editors. 

Thanks for raising this! I've added a comment over on the Gutenberg as Framework issue, too (https://github.com/WordPress/gutenberg/issues/53874#issuecomment-1696615558). Up until this point, I'd assumed that it was desirable for future block supports to be dynamically output via the style engine in both JS and PHP for the reasons discussed back in https://github.com/WordPress/gutenberg/discussions/37495, so I'm keen to better understand if there's been a shift in direction there (or if I've misunderstood, of course 🙂).

In the case of this PR, the main reasons for dynamic output (even of static styles) are:

* To support filterable / dynamic images in follow-ups. Examples include potentially retrieving image url via id, or switching to a featured image, or one day images that form part a theme's assets.
* To allow themes or plugins to filter or switch off output, without breaking markup.
* To allow flexibility in changing output in follow-ups without requiring deprecations.
* To ensure that dependent static rules are only output when needed (so in this case `background-size` is only output when `background-image` is output).

> It would be good to keep that in mind and maybe keep track of all these static/dynamic styles to implement the wp.blocks.render function properly.

Yes, it'd be good to ensure all features can be rendered in the same way in JS or PHP contexts. For this one, since we're adding the logic to the JS version of the style engine, depending on how we might want dynamic JS injection to work in the future (possibly at block render time?), I _think_ overall we're heading in the right direction for that? Please do let me know if I'm missing something though, as I might have missed some of the context behind the thinking in #53874.

ramonjd commented on packages/style-engine/src/styles/index.ts:
```diff
@@ -17,4 +18,5 @@ export const styleDefinitions = [
 	...spacing,
 	...typography,
 	...shadow,
+	...media,
```
At first I questioned the `media` top-level grouping, wondering if we could use `background`, given it has a lot of [related properties](https://developer.mozilla.org/en-US/docs/Web/CSS/background). 

I think it's a good decision however since most, probably all, of those properties manipulate a background element most likely to be an image, and then we'd also have a neat home for stuff like [object-fit](https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit) and anything related to [replaced elements](https://developer.mozilla.org/en-US/docs/Web/CSS/Replaced_element) (videos, canvas, etc)

ramonjd commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -300,6 +300,10 @@ const BlockInspectorSingleBlock = ( { clientId, blockName } ) => {
 					<InspectorControls.Slot group="styles" />
+					<InspectorControls.Slot
+						group="media"
+						label={ __( 'Media' ) }
```
This is most likely overthinking it, but I was wondering if this might be confused with the other "Media", the library? Since it's specific to backgrounds for now, what about "Background media"?

andrewserong replied:
Good question! For the panel title, I was thinking of trying to keep it generic to `Media` for consistency with the Site Logo block that already has a Media panel, and leave the specifics to the individual block support within it. The idea being that we then have a consistent panel across any block that has media, irrespective of where that media winds up being output.

That was just my initial thinking though, happy to change things around! For now, I might leave it as-is while we await further design feedback 🙂

ramonjd replied:
Sounds good to me

ramonjd reviewed (commented):
Thanks for getting this up to test. I reckon it'll open up a bunch of pattern possibilities!

Behold my creation!


https://github.com/WordPress/gutenberg/assets/6458278/38f0db9e-1131-43aa-8628-7b38884bc011

andrewserong committed:
Try using style engine for PHP output, too

andrewserong commented:
Thanks for taking this for a spin @ramonjd! I think all the pieces I intended to include for an MVP are in place now. I'm sure there'll be plenty of further things to tweak, but for now I've switched this over to "ready for review" as a proposal for how the block support could work. If this approach winds up being viable, I'm happy to split it up into smaller changes if it makes things easier for reviewing to land it in an iterative approach, too.

ramonjd commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -535,6 +550,41 @@ protected static function get_individual_property_css_declarations( $style_value
+			if ( ! empty( $style_value['url'] ) ) {
+				$value = "url('" . $style_value['url'] . "')";
+			} elseif ( is_string( $style_value ) ) {
+				$value = $style_value;
```
I like! 

So this could handle both of these scenarios?

```php
$media_block_styles['backgroundImage']['url'] = 'http://some/image.gif';

// or

$media_block_styles['backgroundImage'] = 'url(http://some/image.gif)';


gutenberg_style_engine_get_styles( array( 'media' => $media_block_styles ) 
```

I'm thinking of cases where it might be useful to extend the `background-image` value, e.g., 

```css
background-image: url(http://some/image.gif),
                  url(http://some/image2.gif);
```

andrewserong replied:
Yep, that's what I had in mind! We'd also support `linear-gradient` and other values for this property that way, too.

In future follow-ups it'd be fun to see how we could have layers of each of the different background properties, so you might have an array of background styles and then the style engine could figure out how to output them all in the right sequence (e.g. `background-image` as in your example, with corresponding `background-position` values, for example).

draganescu commented:
> The intention wasn't for media === background media, but rather to create a Media block support, and for the first feature within it, to add a background image support. In follow-ups, we might create all kinds of other media controls,

What other media controls would there be?

In the site logo block the media control is _actually_ adding media to the block. The block then works with it to use it as it has to. This support is basically support for image backgrounds across blocks. Isn't this a style option? When I think "Add media" - I think of media as content, compared to media for styline purposes.

Even the tab on which this is set points to something being off -> I am in the settings tab and seup the background style. I think that we should not add a `media` support but a `background-image` one only. We _can_ use the media control if the block supports background image.

> Block Supports is the API that allows a block to declare support for certain features.

`Media` as "support" is not a feature. Background image is a feature and the block says:" if I am rendered with a background image I'll still, likely, look good" (unlike a paragraph, which does not "support" it). When the block says "I support media" it means pretty much nothing? Now that I think of it even my `supports` or icon fell into this trap 😆 

So I think we should:

- register background-image support
- drop the media support
- use the media control in the styles panel to set background

ramonjd commented:
> Media as "support" is not a feature. Background image is a feature and the block says:" if I am rendered with a background image I'll still, likely, look good" (unlike a paragraph, which does not "support" it). When the block says "I support media" it means pretty much nothing?

Media as a term comes with a lot of baggage and potential confusion, I agree. Let's say we did have "media" supports: what else would go in there? It could be a convenient bucket to house other supports, for example, if we wanted to add video background and object-fit features. 

It's difficult trying to come up with a block support category to house related properties that we conceivably might use in the future. Typography and dimensions etc are intuitive I think and general enough to be welcoming to new members

`background` itself would have been quite fitting, but now not perfect as `background-color` has been already snatched up by color supports 😅 .

 Having said that, `background-image` has many siblings and it wouldn't be such a stretch of the imagination to group `background-repeat` and [others](https://developer.mozilla.org/en-US/docs/Web/CSS/background-position) in there. As I see it, `background-color` is the only one in that family that _doesn't_ affect the style of a background image, so I wouldn't be against it in principle. It's easy to add more supports and much harder to wind them back.

andrewserong commented:
Thanks for the discussion folks, lots of good points, particularly surrounding the nuances of thinking of media as content, in comparison to background image styling. To add to the potential challenges here, let's look at part of how the Cover block is currently implemented. In the Block Toolbar the action to add a background image is called "Add media", and the controls related to the background image appears in the settings tab under the panel "Media settings":

| Cover block toolbar | Media settings in sidebar |
| --- | --- |
<img width="569" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/bcf6f4fd-ce35-42b7-b885-66ad61119d2b"> | <img width="150" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/4760b0c1-6fb1-446f-a98b-6950a58dec3a"> |

Whatever we wind up implementing, I like the idea that in the longer term we move toward a consistent UI between these common container blocks, so if we make some decisions here, let's consider how we might roll that out eventually for the cover block, too. For now, I think there are two separate but related questions:

* What do we call the block support, and where is the data stored in block attributes? I quite like the idea of `background` as a catch all, particularly as most of the other features to implement in shorter-term follow-ups are all related to how background images are presented, but `background` also allows for video backgrounds in the future. So, `attributes.style.background.backgroundImage` and `attributes.style.background.backgroundSize` feel like decent places to store data, and in `block.json` `supports.background.backgroundImage` seems pretty good to me, too.
* Where do we place things in the UI for blocks? @draganescu makes good points about moving to the styles tab. @jasmussen do you have any thoughts on this, I see you mentioned in https://github.com/WordPress/gutenberg/issues/14744#issuecomment-1602176480 about the issues of the Cover block's current UI not scaling well.

For the scope of this PR, we likely want to settle on a good name and place for the data first of all, as that would be much harder to change in follow-ups. I'll have a go at moving things into `background` and see how that feels. Happy to keep iterating as we see how it all looks!

paaljoachim commented:
Does it really have to be background?
Brain storming.... thinking broader strokes.
This seems like a general Add Media setting. Background or foreground could be secondary. It could be a stand alone or layered approach. As in...
Layer 1 contains image, video, audio, document.
If it is layered. These items could be on layer 2 (could be called background) and there could be something on top (layer 1). Be it text, a transparant color (similar to overlay in the Cover block) or something else.

Foreground - layer 1. Could be text, color, image, video...
Background - layer 2. Could be text, color, image, video...
Background - layer 3. Something similar....
Almost as if one is dragging submenu navigations up and down. Shifting the focus between layers.

I mention the above as I believe thinking layers would be of benefit. Instead of just as a background.

andrewserong commented:
Layers is an interesting idea! The earlier background image support also linked out to the discussion about layer management for the Cover block over in https://github.com/WordPress/gutenberg/issues/39427, and there were lots of cool ideas and designs there. One possible idea for layers could be that they might be separate blocks that are layered via z-index values, rather than layers within the one block's supports.

For this PR, I'm keen to see how we can introduce background image support in an iterative way. Since the feature was requested a long time ago now, and with earlier explorations that stalled, I'm hoping if we can settle on the simplest-but-useful version of the support, we could add background images to Group blocks as a first step, and look at further features and enhancements in follow-ups (from repeating background images, to layering ideas, and so on).

andrewserong committed:
Try renaming block support and panel to Background

andrewserong commented:
Update: in https://github.com/WordPress/gutenberg/pull/53934/commits/caaa71bcf3a7d93d102901f7da618109c5a63700, I've had a go at renaming the overall block support and panel to Background, retaining `backgroundImage` as the sub-feature. Here's how it's looking now in the UI:

![image](https://github.com/WordPress/gutenberg/assets/14988353/f731888d-f8d7-427b-8513-f7b05ca5b32c)

One drawback already mentioned about a general Background panel is that it doesn't include background color. However, since the panel is now right next to the Color panel, where background color is featured, I wonder if that helps mitigate things?

Thanks again for all the feedback and discussion so far, folks! Happy to try out further renaming and to keep iterating 🙂

apeatling commented:
> One drawback already mentioned about a general Background panel is that it doesn't include background color. 

Background Media?

> I mention the above as I believe thinking layers would be of benefit. Instead of just as a background.

This is a nice idea, but IMO we should keep the scope smaller on this PR so we can get the initial version of background support merged.

ramonjd commented:
> One drawback already mentioned about a general Background panel is that it doesn't include background color. However, since the panel is now right next to the Color panel, where background color is featured, I wonder if that helps mitigate things?

"background" as a block supports group sounds reasonable to me! It's independent of the UI so we could guide users by any copy we like. 

I'd vote for "Background Image" for this PR.

Thanks for sticking with this.

jameskoster commented:
Agree on the control being present in Styles. This also makes it possible to set the background globally which can be handy.

> I'd vote for "Background Image" for this PR.

This seems reasonable to me.

jasmussen commented:
Small preference for just "Media" 😅

jameskoster commented:
"Media" is an existing panel on the Site Logo block (and one you might expect to find on blocks like Image, Video, etc. as well). 

I think the application here is different:

1. It's an optional style rather than a setting that is essentially required for the block to perform its intended function.
2. The media is applied to the background.

This warrants a different name, imo.

jasmussen commented:
Right, good point. Thinking a bit ahead, you might want both an image, and a background color in case it's a transparent PNG, and even background tiling options. So to that end, perhaps "Background" is good? The UI definitely needs further iteration of course, but in terms of keeping it in one panel.

andrewserong committed:
Rename panel label to Background image, rename component for consistency

andrewserong commented:
Good discussion, folks! I do quite like the idea of "Background image" for distinguishing between this new panel and background color as it appears in the color panel (I think we're using sentence case still for these sorts of headings?). Also, it sounds like most of our ideas for follow-up controls will still be related to how background images are displayed, so there's some consistency there. For now, I've renamed the panel to "Background image" so we can see how it feels. Here's a screenshot of the current state of the PR:

<img width="1419" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/5e193ef3-6851-4800-a864-aeeaa13b4ec4">

Happy to keep iterating of course!

jameskoster commented:
Yeah "Background" alone is a little awkward because it kind of suggests you'd find the background _color_ option there too. 

I do wonder if at some point color options should be more contextually placed (background color in "Background" panel, text color in "Typography" panel), but that's totally separate.

andrewserong commented on packages/block-editor/src/hooks/style.js:
```diff
@@ -127,7 +129,7 @@ const skipSerializationPathsEdit = {
 const skipSerializationPathsSave = {
 	...skipSerializationPathsEdit,
-	[ `${ SPACING_SUPPORT_KEY }` ]: [ 'spacing.blockGap' ],
+	[ `${ BACKGROUND_SUPPORT_KEY }` ]: [ BACKGROUND_SUPPORT_KEY ], // Skip serialization of background support in save mode.
 };
```
I'll look into this tomorrow, but I just noticed that if I switch off background image support at the block level, and have a block that already contains the attributes for background image, then the background image winds up being serialized to markup:

<img width="1272" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/4222f2c7-dcbd-494d-b4fb-a9ff29f45939">

In the case of the background image support, we want it to always be stripped from the saved markup, irrespective of whether or not the block supports the value.

andrewserong replied:
This is mostly resolved now, but I'll add a follow-up question below.

andrewserong committed:
Update block.json schema, fix issue with omitting styles

andrewserong commented:
Update: I've updated the `block.json` schema. In the process, I was testing what happens when we switch off the block support. A question for folks:

If you have block markup that already has a background image in it, what do you expect will happen if a theme does not have the background image support enabled? We have a few options:

1. Hide the UI controls for background images, but still render the background image for the block in the editor and on the site frontend (this means that any theme could start using block patterns that contain background images)
2. As above, but _do_ show the UI if a block already has a background image. If a user removes background image from a block, then the controls would become hidden.
3. Hide the UI controls _and_ hide the output of the background image in both the editor and on the site frontend — this would prevent background images from being displayed anywhere, but it would also mean that only themes that opt-in to the background image support could see patterns that contain background images.

I think I'm leaning slightly toward the idea of number (2), still rendering in the editor and site frontend, and expose the controls in such a case so that folks can remove an image if need be, but would love for some feedback on that, as it's tricky to know what will be most logical for folks.

The current state of the PR is sort of stuck in-between, so I'll need to fix this up before the PR is mergeable.

andrewserong commented on lib/block-supports/background.php:
```diff
@@ -0,0 +1,103 @@
+function gutenberg_render_background_support( $block_content, $block ) {
+	$block_type                   = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+	$block_attributes             = $block['attrs'];
+	$has_background_image_support = block_has_support( $block_type, array( 'background' ), false );
```
We could remove this check and replace it with a check for whether a background image is set in the block's attributes, if we'd like to ensure background images will always render, irrespective of whether a theme has support for it.

I'm a bit undecided, though, so happy for feedback here!

ramonjd commented:
> If you have block markup that already has a background image in it, what do you expect will happen if a theme does not have the background image support enabled? We have a few options:

I also like the sound of two.

I had remind myself what we do for other supports. Looks like we turn them off? Here's margin on/off between two themes

| Without margin support | With margin support |
| ------------- | ------------- |
| <img width="400" alt="Screenshot 2023-09-07 at 10 47 42 am" src="https://github.com/WordPress/gutenberg/assets/6458278/8f4cd429-cf8c-415c-84c0-c11a92c75b00">  | <img width="400" alt="Screenshot 2023-09-07 at 10 47 26 am" src="https://github.com/WordPress/gutenberg/assets/6458278/6f913f07-a54a-489d-9cb6-7b99c2c78754">  |

But having the control to remove it in the case of image background would be helpful since background images are a bit more ostentatious. Do you think it would be helpful to warn the user know that the theme doesn't support the feature?

ramonjd commented:
> Hide the UI controls for background images, but still render the background image for the block in the editor and on the site frontend (this means that any theme could start using block patterns that contain background images)

Also, if this is what we do right now with other supports, I'd also add that it'd fine to postpone the decision for another PR 😄

andrewserong committed:
Simplify block schema slightly, make the check for the background.backgroundImage property

andrewserong commented:
Thanks for the feedback, Ramon!

> Also, if this is what we do right now with other supports, I'd also add that it'd fine to postpone the decision for another PR 😄

Thanks for confirming — I think we're in a pretty good place with the support now, then. I've updated the block.json schema slightly so that it refers to `background` as an object only, as I noticed the appearance tools opt-ins make it challenging to get the root value working as a boolean. I think that'll be another one to defer to other PRs.

All up, I think this PR is probably in a good place for either final review or more design feedback now. The MVP-like functionality is all in place, and the background image will still gracefully fall back to rendering, with the UI hidden, if the background image exists in block attributes. This helps ensure background images work well in patterns, and we can look at exposing the UI in themes that don't formally support the control in follow-ups, as you mention.

Let me know if anyone has any other feedback they'd like explored at this stage!

jameskoster commented:
Agree option number two sounds better than one and three. It's not perfect though – I could definitely see users reseting the background image then not understanding why they can't upload a new one. I'm happy to acknowledge that's an edge case. 

One tiny detail for consideration; in the control, did y'all consider displaying the image Title (if present) rather than the file name? I'm wondering if that might be a little more intuitive. 

<img width="299" alt="Screenshot 2023-09-07 at 09 21 04" src="https://github.com/WordPress/gutenberg/assets/846565/897f4db1-d091-430f-9582-a7c430c1927f">
<img width="281" alt="Screenshot 2023-09-07 at 09 21 16" src="https://github.com/WordPress/gutenberg/assets/846565/942ef9bf-8e80-42ff-b2aa-df300256ccf2">

draganescu commented on packages/block-editor/src/hooks/background.js:
```diff
@@ -0,0 +1,271 @@
+		return true;
+	}
+
+	return !! support?.[ feature ];
```
Should we treat the case of feature === any? I see in other places `packages/block-editor/src/hooks/border.js` it is treated.

andrewserong replied:
Good catch, yes, I'd removed the check because we only have one feature for now, but we should still handle the `any` case. I've added a check back in, which gives us a natural place to put other support features in follow-ups when we add in additional controls 👍

draganescu commented on packages/block-editor/src/hooks/background.js:
```diff
@@ -0,0 +1,271 @@
+		}
+
+		if ( isBlobURL( media.url ) ) {
+			// TODO: Might need to revoke the blob URL.
```
what does this mean?

andrewserong replied:
Oh, thanks for the reminder! I needed to double check — when media is uploaded a temporary blob URL is created and added to a cache. It turns out it looks like the revoking / clearing of the blob URL is already handled for us (if I'm following correctly) by the upload logic [here](https://github.com/WordPress/gutenberg/blob/53a44bedfc90451592c895e245bc662f3b345333/packages/media-utils/src/utils/upload-media.js#L69) once a file has been uploaded, so we don't need to handle it in this component. I've removed the TODO comment.

draganescu commented on packages/block-editor/src/hooks/background.js:
```diff
@@ -0,0 +1,271 @@
+			resetAllFilter={ resetAllFilter }
+			panelId={ clientId }
+		>
+			<div className="block-editor-hooks__background__inspector-media-replace-container">
```
I expected this to be a reused component with the site logo. I guess it's easy to build, but we may end up with different looking controls in time.

andrewserong replied:
Yes, I thought we could look at consolidating the components once this PR lands and we feel confident about the UI being consistent. I figured it'd be a bit easier to keep them separate while we introduce the new support, and then could look at building a shared component in a follow-up.

draganescu commented on phpunit/style-engine/style-engine-test.php:
```diff
@@ -493,6 +493,25 @@ public function data_wp_style_engine_get_styles() {
+				'expected_output' => array(
+					'css'          => "background-image:url('http://example.com/image.jpg');background-size:cover;",
+					'declarations' => array(
+						'background-image' => "url('http://example.com/image.jpg')",
```
I wonder if we should use `https` in the examples 😀

andrewserong replied:
Good call! Done.

draganescu reviewed (approved):
This is a great addition. To me it tested perfectly and I love the UX this offers now. Went through the code and left a few minor comments. 

Excellent progress 👏🏻 on a much awaited feature 🎉

ajlende commented on packages/block-editor/src/hooks/style.js:
```diff
@@ -127,7 +129,11 @@ const skipSerializationPathsEdit = {
  */
 const skipSerializationPathsSave = {
 	...skipSerializationPathsEdit,
-	[ `${ SPACING_SUPPORT_KEY }` ]: [ 'spacing.blockGap' ],
```
Should this have been removed?

andrewserong replied:
Ah, apologies, I should have left a comment about this! Ever since #40875 where `--wp--style--block-gap` was removed from `blocks/src/api/constants.js`, the `spacing.blockGap` value hasn't been handled by inline styles, so this line wasn't doing anything. Since we're adding something to this object for this PR, I thought it'd be a good time to tidy up and remove this other line that's no longer needed.

ajlende commented:
I see there was already some serialization discussion in https://github.com/WordPress/gutenberg/pull/53934#issuecomment-1695271930. But I wanted to add that another benefit of skipping serialization (at least for now) is that we can easily swap out the `background-image` CSS for an `img` element that supports `srcset` and SVG filters.

I think `srcset`, in particular, is rather important for site performance since cover backgrounds can be quite large and you don't usually want to load the full resolution on mobile. Unfortunately, you can't always use an `img` element because it is incompatible with `background-repeat`. (Eventually we'll have [`image-set`](https://caniuse.com/css-image-set) in CSS, but it isn't widely supported yet, and where it is it often needs the `-webkit-` prefix still.)

Second, folks will probably start asking for duotone support once the group supports background images. A separate background `img` element will also be required in that case so the filter isn't applied to the content.

Like I said, since we aren't serializing anything, these changes will be easy to make in follow-ups. They don't need to block this PR. Just thought it was worth mentioning.

ramonjd commented:
Took it for another test run. 

LGTM in all browsers

Image displays correctly with `background.backgroundImage` support turned off in block.json and theme.json (and in classic themes) 👍🏻 

Question: do we need to also update the theme.json schema?

andrewserong committed:
Update test, remove TODO comment

andrewserong committed:
Fix upload container classname

andrewserong committed:
Update hasBackgroundSupport to handle any case

andrewserong committed:
Update theme.json schema

andrewserong committed:
Store media title in attributes, update docs

andrewserong commented:
Thanks again for the feedback, everyone! I think I've implemented or replied to everything. Main changes:

* Updated theme.json schema
* Updated tests to use `https` example link
* Updated hasBackgroundSupport to handle `any` case
* Updated `mediaSelect` to store the title of the media and use this as the label

Here's how it's looking now with the Title label:

https://github.com/WordPress/gutenberg/assets/14988353/4257ff41-1929-4f18-b4a4-f283fa1601c4

I'll wait until next week for the final merge — assuming designers are happy with this for an MVP, I think we might be just about ready to land it now 🙂

jameskoster reviewed (approved):
This looks good to me as a 'v1'. 

I think the control might benefit from a quick way to unset the current image, but that's not something that needs to be handled here.

andrewserong commented:
Thank you for all the collaboration here, everyone! 🙇

andrewserong commented:
I've opened up a follow-up issue for ideas for future enhancements over in #54336. Feel free to add any ideas there! I imagine most enhancements other than #54319 (add Reset option) will likely be for beyond 6.4.

andrewserong commented:
I've opened a backport PR for the PHP parts of this feature over in https://github.com/WordPress/wordpress-develop/pull/5209

andrewserong commented:
Related to the backport PR, I've opened up a separate Gutenberg PR to add in some more PHP tests over in https://github.com/WordPress/gutenberg/pull/54489

westonruter commented:
Re: https://github.com/WordPress/gutenberg/pull/53934#issuecomment-1710478769

> But I wanted to add that another benefit of skipping serialization (at least for now) is that we can easily swap out the `background-image` CSS for an `img` element that supports `srcset` and SVG filters.

@ajlende Another benefit of an `img` element is that WordPress can optimize its loading by adding the `fetchpriority="high"` attribute. WordPress doesn't have any such optimizations currently for `background-image`, although this is something I'm exploring.

Example comparing background images in Group block versus Cover block, where only latter gets `fetchpriority=high` correctly: https://gist.github.com/westonruter/107f0c827bbee12341441c527ea2ed70

andrewserong commented:
Draft dev note:

---

## Background image block support

In WordPress 6.4, a new background image block support has been added, with the Group block opted-in by default. For themes using the `appearanceTools` feature in `theme.json`, the control will be available in the inspector controls, under Background.

The feature allows users to set a background image for a Group block. The selected image will be output at render-time for the block via an inline style applied to the block's wrapper. When a background image is set, the rule `background-size: cover` is output along with the `background-image` property. This ensures that any background images adequately cover the block. Ideas for enhancements in future releases are being tracked in this Github issue: https://github.com/WordPress/gutenberg/issues/54336
### How to add backgroundImage support to a theme

There are two ways to add support for `backgroundImage` to a block theme. The simplest is to opt in to the `appearanceTools` setting, which automatically enables a number of design tools ([read more in the developer handbook](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/#opt-in-into-ui-controls)).

For themes that wish to have more granular control over which UI tools are enabled, the `backgroundImage` support can be opted into by setting `settings.background.backgroundImage` to `true` in `theme.json`. For example:

```
{
	"settings": {
		"background": {
			"backgroundImage": true
```

Note that as of WP 6.4, the `backgroundImage` support is only available at the individual block level, and not in global styles or at the site root. These features will be explored in subsequent releases, with progress tracked in this Github issue: https://github.com/WordPress/gutenberg/issues/54336.

ramonjd commented:
> This is I think not the first place where we favor server rendering over post content serialization for styles. The downside of this approach is that it breaks these blocks for third-party block editors. 

> In the case of this PR, the main reasons for dynamic output (even of static styles) are:
> 
> To allow flexibility in changing output in follow-ups without requiring deprecations.


I'm looking at trying to prevent `color.gradient`, which outputs its value using the `background` shorthand CSS property, from overwriting all `background-*` CSS properties in https://github.com/WordPress/gutenberg/pull/60739. 

Gutenberg has backed itself into a pretty tight corner by using `background` in this way. Ideally `background-image` should be used for images and linear gradients and combined where both exist.

Because we're skipping serialization for background block supports however, to achieve the above means that some
complicated deduplication and string manipulation needs to occur on the server where a block has both a gradient and background image:

1. When the incoming block has a background gradient applied in the inline CSS - here I want to remove that CSS and use `background-image` only.
2. Where a background gradient preset is applied via classname (which all have !important in the corresponding CSS rule) - the classname should be removed in favour of a combined `background-image` value.

Just flagging that changing output is not as flexible or convenient as it one might assume.

> To ensure that dependent static rules are only output when needed (so in this case background-size is only output when background-image is output).

Relative to background-image, is this such a bad thing? If an image has been reset in the editor, Gutenberg can remove the style properties at that point. If background-size is set elsewhere, e.g., in theme.json and there's not image, I don't see why Gutenberg should care.
